eliot/_action.py

```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
+            C{FAILED_STATUS}.
+        @raise InvalidStartMessage: If C{start_message} does not have a
+            C{ACTION_STATUS_FIELD} of C{STARTED_STATUS}, or if it has a
+            C{task_level} indicating that it is not the first message of an
+            action.
+
+        @return: A new, completed C{WrittenAction}.
+
+        @return: A new C{WrittenAction}.
+        """
+        if start_message.contents.get(ACTION_STATUS_FIELD, None) != STARTED_STATUS:
+            raise InvalidStartMessage.wrong_status(start_message)
+        if start_message.task_level.level[-1] != 1:
+            raise InvalidStartMessage.wrong_task_level(start_message)
+        action_type = start_message.contents.get(ACTION_TYPE_FIELD, None)
+        status = STARTED_STATUS
+        action = cls(
+            action_type=action_type,
+            status=status,
+            task_uuid=start_message.task_uuid,
+            task_level=start_message.task_level.parent(),
+            start_time=start_message.timestamp,
+            end_time=None,
+            _children=m(),
+            exception=None,
+            reason=None,
+        )
+        for child in children:
+            action = action._add_child(child)
+        if end_message:
+            return action._end(end_message)
+        return action
+
+    @property
+    def children(self):
+        """
+        The list of child messages and actions, excluding the start and end
+        messages.
+        """
+        return self._children.values()
+
+    def _validate_message(self, message):
+        """
+        Is C{message} a valid direct child of this action?
+
+        @param message: Either a C{WrittenAction} or a C{WrittenMessage}.
+
+        @raise WrongTask: If C{message} has a C{task_uuid} that differs from the
+            action's C{task_uuid}.
+        @raise WrongTaskLevel: If C{message} has a C{task_level} that means
+            it's not a direct child.
+        """
+        if message.task_uuid != self.task_uuid:
+            raise WrongTask(self, message)
+        if not message.task_level.parent() == self.task_level:
+            raise WrongTaskLevel(self, message)
+
+    def _add_child(self, message):
+        """
+        Return a new action with C{message} added as a child.
+
+        Assumes C{message} is not an end message.
+
+        @param message: Either a C{WrittenAction} or a C{WrittenMessage}.
+
+        @raise WrongTask: If C{message} has a C{task_uuid} that differs from the
+            action's C{task_uuid}.
+        @raise WrongTaskLevel: If C{message} has a C{task_level} that means
+            it's not a direct child.
+
+        @return: A new C{WrittenAction}.
+        """
+        self._validate_message(message)
+        level = message.task_level
+        if self._children.get(level, message) != message:
+            raise DuplicateChild(self, message)
+        return self.set(_children=self._children.set(level, message))
+
+    def _end(self, end_message):
```

Comment:
The problem with this particular mechanism is that it loses any additional fields the end message may have. Success messages may have additional fields, at least. And now that I think of it there's the same problem with any extra messages the start message has.

So perhaps both start and end message should be attributes of the `WrittenAction`, and some of the things that are currently attributes become properties?